pkg/controller/hostpathprovisioner/daemonset.go

```diff
@@ -28,15 +28,33 @@ import (
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/client-go/tools/record"
 	hostpathprovisionerv1 "kubevirt.io/hostpath-provisioner-operator/pkg/apis/hostpathprovisioner/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
 // reconcileDaemonSet Reconciles the daemon set.
 func (r *ReconcileHostPathProvisioner) reconcileDaemonSet(reqLogger logr.Logger, cr *hostpathprovisionerv1.HostPathProvisioner, namespace string, recorder record.EventRecorder) (reconcile.Result, error) {
+	// Previous versions created resources with names that depend on the CR, whereas now, we have fixed names for those.
+	// We will remove those and have the next loop create the resources with fixed names so we don't end up with two sets of hpp resources.
+	if cr.Name != MultiPurposeHostPathProvisionerName {
+		dup, err := r.checkDuplicateDaemonSet(cr.Name, namespace)
```

Comment:
Okay so the pattern I would use for this is the following:
```go
  dups, err := r.getDuplicateDaemonSet()
  if err != nil {
    return reconcile.Result{}, err
  }
  for dup, _ := range dups {
    if err := r.deleteDaemonSet(cr.Name, namespace); err != nil {
      return reconcile.Result{}, err
    }
  }  
```
No need to log the error, when we return the err as part of the result it will get logged. And have getDups return a list of daemonsets:
```go
func (r *ReconcileHostPathProvisioner) getDuplicateDaemonSet() ([]appsv1.DaemonSet, err) {
  dsList := &appsv1.DaemonSetList{}
  dups := make(([]appsv1.DaemonSet, 0
  ls, err := labels.Parse(fmt.Sprintf("k8s-app=%s", MultiPurposeHostPathProvisionerName))
  if err != nil {
     return dups, err
  }
  if err := r.client.List(context.TODO(), dsList, &client.ListOptions{LabelSelector: ls, Namespace: namespace}); err != nil  {
    return dups, err
  }
  for _, ds := range dsList.Items {
    if ds.Name != MultiPurposeHostPathProvisionerName {
      dups = append(dups, ds)
    }
  }
  return dups, nil
}
```
This way we delete any daemonsets with the label that don't match our new value, and return to the required state.

Reply:
Done